cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramNotificationSubscriberService.java

```diff
@@ -396,7 +431,36 @@ private void writeToHeartBeatDataset(@Nullable RunRecordMeta recordedRunRecord,
         appMetadataStore.recordProgramDeprovisioning(programRunId, messageIdBytes);
         return Optional.of(provisioningService.deprovision(programRunId, datasetContext));
       case DEPROVISIONED:
-        appMetadataStore.recordProgramDeprovisioned(programRunId, endTs, messageIdBytes);
+        RunRecordMeta meta = appMetadataStore.recordProgramDeprovisioned(programRunId, endTs, messageIdBytes);
+        if (meta == null || meta.getCluster() == null) {
+          // this should not happen since when the cluster info should be there after deprovision
+          return Optional.empty();
+        }
+        Optional<ProfileId> profile = SystemArguments.getProfileIdFromArgs(programRunId.getNamespaceId(),
+                                                                           meta.getSystemArgs());
+
+        if (!profile.isPresent()) {
+          // Each program run should be associated with a profile, so this should never happen
+          LOG.warn("Will not emit metrics for program {} without profile assigned to it, {}",
+                   programRunId, notification);
+          return Optional.empty();
+        }
+
+        ProfileId profileId = profile.get();
+
+        // emit the metrics information, increment the count for end state program runs
```

Comment:
its better to emit these metrics when we handle the notifications for killed/completed/failed for program runs rather than here for deprovisioned. You wouldn't need the getMetricName method anymore, and the code would be easier to follow.

We also need to be careful not to emit repeat metrics for those in case processing these notifications are retried due to failure. I think it's better to return an optional of a runnable that increments the metric so that it gets incremented only after all these notifications are successfully processed in a transaction.

Reply:
Moved the logic to handleProgramEvent, and also make that method return a Runnable